### ISSUE TITLE
Searchable documentation?

### ISSUE THREAD

**Issue by ivirshup** [bokeh/bokeh#8209]
There used to be a search bar in the documentation, which now appears to have been removed. For me, this makes the documentation site a lot less helpful.

<img width="698" alt="screen shot 2018-09-04 at 1 26 55 pm" src="https://user-images.githubusercontent.com/8238804/45008562-b2dfa280-b046-11e8-9201-065817ed7f38.png">
<img width="702" alt="screen shot 2018-09-04 at 1 27 06 pm" src="https://user-images.githubusercontent.com/8238804/45008566-b4a96600-b046-11e8-9bcf-bbbaf8a41614.png">

Is this something that's coming back?

**Comment by bryevdv** [comment]
Unfortunately, Google shut down "Google Site Search" which powered the old search box (see https://github.com/bokeh/bokeh/issues/5911 for their email) I'd love to have a useful (ad-free) search box. But I don't know how to make that happen.  Getting it back will require someone with relevant expertise to decide to step in and help lead the effort. 

Bing is a possible alternative:

https://blogs.bing.com/2017-05/build-your-own-web-search-service-with-bing-custom-search
https://azure.microsoft.com/en-us/pricing/details/cognitive-services/bing-custom-search/

We would have to find a way to fund a budget for using it though NunFocus, unless there are any free offerings for OSS projects. 

Alternatively, the docs are actually hosted on a server. If there is some indexing engine that we can run ourselves, that might be an option. Again, anyone with expertise would be extremely appreciated to come help make this happen.

**Comment by mattpap** [comment]
Given that bokeh's docs are indexed by google search nevertheless, as a temporary measure, one can use queries like `site:bokeh.pydata.org gmapplot` to limit search only to bokeh's website. Also, is there a reason why we don't just use sphinx's (client side) search functionality?

**Comment by bryevdv** [comment]
> Also, is there a reason why we don't just use sphinx's (client side) search functionality?

It's not very good, and requires dumping the entire index into a huge JS blob on every page, IIRC. But maybe things are better, it could be worth a try, too.

**Comment by ivirshup** [comment]
Out of curiosity, how are the datashader, pyviz, and holoviews sites doing their search?

**Comment by jbednar** [comment]
We copied the site configuration for all those sites from Bokeh originally, so alas, we have had the same problem Bokeh has; the search now just returns bogus full-web search results. :-(  

We'd love to have something set up that's more useful, even if it's just a specially crafted URL for google to focus the search to our site.

**Comment by hklarner** [comment]
A message explaining the disappearance or a link to this issue in place of the search box would prevent annoyed bewilderment. Its just very uncommon to offer a documentation that is not searchable.

**Comment by ivirshup** [comment]
Would it be possible to just have the search route you to something like: `https://www.google.com/search?site:bokeh.pydata.org+inurl:{current_version}+{query}`?

**Comment by bryevdv** [comment]
I didn't have any luck:

<img width="764" alt="screen shot 2019-02-11 at 23 36 20" src="https://user-images.githubusercontent.com/1078448/52619190-e294a100-2e55-11e9-889f-d01664ccef79.png">

I also tried without the `{}`, and with double quotes around the text. Google does not seem to like the `.'` characters. At one point it also said I was sending suspicious traffic and made me pass a CAPTCHA. So, I can;'t say that's a promising start. But it gets to my earlier point:

> Getting it back will require someone with relevant expertise to decide to step in and help lead the effort.

OSS Is a collaboration. if someone with the right experience (or time) can prove this out, I'd love to try it.

**Comment by ivirshup** [comment]
Oh, that was kind of a rough f-string style template. A little more formally (using python conventions) it'd be:

```python
f"https://www.google.com/search?q=site:bokeh.pydata.org+inurl:{version_in_url}+{query.replace(' ', '+')}"
```

For example, searching for "network plot" on the latest docs would send you to [https://www.google.com/search?q=site:bokeh.pydata.org+inurl:latest+network+plot]()

**Comment by bryevdv** [comment]
@ivirshup Ah, until we ditch Python 2 later this year, f-strings are not usable by us so I have not really used them either. 

Unfortunately, while that works for "latest" (which was also true of what I tried earlier) it does *not* eek for "1.0.4" again google really seems to have the dots

<img width="809" alt="screen shot 2019-02-12 at 13 26 28" src="https://user-images.githubusercontent.com/1078448/52669048-07c2f700-2eca-11e9-94a5-cd36949b32d0.png">

**Comment by ivirshup** [comment]
Ah yeah, I had only checked if it worked for `inurl:latest`. Anyways, I'd assumed this would get implemented in javascript instead of python, since it only need information from the current page.

So I guess like:

```javascript
function search(query) {
    ver = window._BOKEH_CURRENT_VERSION
    url = `https://www.google.com/search?q=site:bokeh.pydata.org+${ver}+${query.replace(/ /g, "+")}`
    window.open(url, "_self")
}
```

So `search("network plots interactive")` would take you to: https://www.google.com/search?q=site:bokeh.pydata.org+1.0.4+network+plots+interactive

But maybe with some URI cleaning?

**Comment by bryevdv** [comment]
@ivirshup That's promising! I will see if I am able to incorporate something before the next release

**Comment by bryevdv** [comment]
At this point I think I will give the sphinx search plugin another shot. It's been several years, perhaps things have improved somewhat.

**Comment by bryevdv** [comment]
Another option would be to try and publish to RTD, since they have their own search indexing AFAIK,  but that seems like a huge undertaking. Our docs build is very complicated and seems to not fit neatly into their expectations.

**Comment by bryevdv** [comment]
So, to give an update. I am going to work towards migrating our docs to RTD. Even possibly engaging with them for commercial consulting on the project. I can't speculate how long this might take but my hope is sometime in the next few months.

**Comment by bryevdv** [comment]
Further update: we have submitted a NumFocus small project grant proposal to fund paying RTD to implement a migration. failing that, there are some other funds we can allocate. Best hope for completion is mid 2019

**Comment by blaise-sumo** [comment]
@bryevdv The RTD folks are great, and IIRC they have a special deal for open source projects on their commercial site readthedocs`.com`

Thanks for doing so much for the docs.

> Further update: we have submitted a NumFocus small project grant proposal to fund paying RTD to implement a migration. failing that, there are some other funds we can allocate. Best hope for completion is mid 2019

**Comment by bryevdv** [comment]
Unfortunately despite everyone's best intentions, the RTD integration did not succeed, for a variety of reasons. Our intent is now to publish the docs as a static site behind CloudFront, and integrate a FTS service such as [algolia](https://www.algolia.com) at their free or cheap tier.

**Comment by ivirshup** [comment]
Thanks for the update!

I noticed there was a search tab on the docs site over the past week or so that seems to be gone now. Any chance that could stick around until a more permanent solution is found?

**Comment by bryevdv** [comment]
It was never actually active on the real deployed site, only on the test deploys to RTD. In any case, I have had to revert all that work. That search is 100% decedent on RTD builds which we are not using.

**Comment by bryevdv** [comment]
OK I was feeling pretty despondent about this, all the fully managed SaaS search offerings are way too expense (starting at 50 USD/mo and going up from there). Client search (e.g. lunr.js, etc) would be slow and require an enormous JS index file load. Running your own FTS is both super complicated, and still fairly more expensive than we can afford (might raise our current AWS spend by 50% or more)

There is a free service called Doc Search provided by Algolia. The default (fairly unconfigurable) result seemed really subpar to me, only showinfg up to 5 results in an instant search box. But I finally stumbled across this fiddle (created by one of the Algolia devs) that shows how to use their index in a more sophisticated way to get a much larger paginated set of results:

https://jsfiddle.net/maxiloc/oemnhuv4/

I am waiting to get the project enrolled in DocSearch to test this out, but fingers crossed...

**Comment by bryevdv** [comment]
Very happy news, got the necessary information back and plugged it in and it seems to work just fine. Here is the same fiddle above tailored to Bokeh's new index:

https://jsfiddle.net/xk1hd4qg/

I will proceed on this basis!

**Comment by bryevdv** [comment]
I'm afraid I have to tamp down expectations here once again. The default configured "public docsearch" is very poor. I had thought I could improve it as described above but the algolia staff seems to b e frustrated by my asks. They told me to "run the scraper myself first" to sort out the config,but apparenlty just doing that ***and nothing else*** was enough to generate "64000 operations" (whatever those are) which is enough overbudget that the account was suspended. If they merge the latest PR then perhaps things will work out. If not, this was probably another dead end. 

In that case, I guess realistically the only option (and the cheapest workable one I know of) woudl be to run our own FTS and server but I'm not sure confident have the budget for it. A DO box powerful enough might push 20 USD/mo

**Comment by ivirshup** [comment]
@bryevdv, just another possibility I stumbled on, [google's current custom site search](https://cse.google.com/?ref=b&hl=en) seems to work alright. [Here's an example](https://cse.google.com/cse?cx=001886091806704350059:ogav6lff3dl) pointed at the latest docs. There are [free plans](https://support.google.com/customsearch/answer/9069107?hl=en&ref_topic=4513742) (where it runs in the user's browser) and looks like results can be [customized](https://developers.google.com/custom-search/docs/annotations) – which would be nice for giving reference docs higher priority.

**Comment by bryevdv** [comment]
@ivirshup Thank you for the suggestion, but unless something has changed that I don't know about, GCSE requires admitting ads, which I am not willing to do.

**Comment by ivirshup** [comment]
I was thinking since the docs are hosted under pydata, maybe that could count for the non-profit tier.

> runs in the user's browser ... What exactly does this mean? 

This is a little unclear to me as well. There's some mention of it being javascript based, but the page isn't too huge. I think they might build an index on their servers and ship that to the client?

> I personally would rather have narrative docs show up first.

Definitely agree with this, but it would be nice to get the reference docs for `HoverTool` in the first page if I search for `hover`. I think pages with the search term in a section heading could use a boost in the rankings.

> I also would really like a solution that affords faceting by release version

This looks possible. The example I sent is limited to `https://bokeh.pydata.org/en/latest/`.

**Comment by bryevdv** [comment]
@mattpap @bokeh/core FYI NumFOCUS is working on registering the appropriate non-profit status with Google, at which point we *should* be able to create a CSE for *docs.bokeh.org* with ads disabled. It may take another week or so to complete. I'd propose to stall 1.4 release until week after next (if necessary) to try and get a search into the 1.4 docs.

**Comment by mattpap** [comment]
> I'd propose to stall 1.4 release until week after next (if necessary) (...)

That's fine. Though it would be good to have the release done before November 2nd.

**Comment by bryevdv** [comment]
> That's fine. Though it would be good to have the release done before November 2nd.

Agree 100%, if the search is not available by then I would intend to just release 1.4 anyway

**Comment by bryevdv** [comment]
For now got the client-side sphinx search working again. Will switch to ad-free GCSE under NF status, if we are able.
